Encryption in Mongodb

phiz71 opened the pull request:
## Issue

N/A

## Description

Enable encryption feature with mongodb and allow encryption of some sensitive data in User and Metadata
<!-- Environment placeholder -->

🏗️ Your changes can be tested here and will be available soon:
      Console: [https://pr.team-apim.gravitee.dev/10575/console](https://pr.team-apim.gravitee.dev/10575/console)
      Portal: [https://pr.team-apim.gravitee.dev/10575/portal](https://pr.team-apim.gravitee.dev/10575/portal)
      Management-api: [https://pr.team-apim.gravitee.dev/10575/api/management](https://pr.team-apim.gravitee.dev/10575/api/management)
      Gateway v4: [https://pr.team-apim.gravitee.dev/10575](https://pr.team-apim.gravitee.dev/10575)
      Gateway v3: [https://pr.gateway-v3.team-apim.gravitee.dev/10575](https://pr.gateway-v3.team-apim.gravitee.dev/10575)

<!-- Environment placeholder end -->
<!-- Storybook placeholder -->
---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-htxmwxnlky.chromatic.com)
<!-- Storybook placeholder end -->

phiz71 committed:
refactor: adapt spring structure of some classes

Allow to add new MongoConverters more easily

phiz71 committed:
feat: add support for mongodb encryption

Allow to enable encryption mode in mongodb

phiz71 committed:
test: avoid running default-test twice

phiz71 committed:
feat: add mongodb encryption on user

phiz71 committed:
feat: add mongodb encryption on metadata

jhaeyaert commented on gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/encryption/EncryptionConfiguration.java:
```diff
@@ -0,0 +1,162 @@
+    private void initDataEncryptionKey(ClientEncryption clientEncryption) {
+        BsonDocument keyByAltName = clientEncryption.getKeyByAltName(keyAlternativeName);
+        if (keyByAltName == null) {
+            clientEncryption.createDataKey(KEY_MANAGEMENT_SYSTEM_LOCAL_TYPE, new DataKeyOptions().keyAltNames(List.of(keyAlternativeName)));
```
For the record, like we talked about, we can add a try/catch clause to catch the unique violation exception in case two instances try to create the DEK at the same time.

jhaeyaert commented on gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/encryption/EncryptionConfiguration.java:
```diff
@@ -0,0 +1,162 @@
+        ClientEncryptionSettings.Builder builder = ClientEncryptionSettings
+            .builder()
+            // The collection in MongoDB where the Data Encryption Keys (DEKs) will be stored.
+            .keyVaultNamespace(getDatabaseName() + "." + tablePrefix + keyVaultCollectionName)
```
We can try removing the getDatabaseName() and see if it reuses the same database from the mongoFactory.

phiz71 replied:
It works without the database name. I'll remove that part.

jhaeyaert commented on gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/MongoUserRepository.java:
```diff
@@ -46,18 +46,28 @@ public class MongoUserRepository implements UserRepository {
-        UserMongo user = internalUserRepo.findBySourceAndSourceId(source, escapedSourceId, organizationId);
+        UserMongo user;
+        if (isEncryptionEnabled) {
+            user = internalUserRepo.findBySourceAndSourceId(source, sourceId.toLowerCase(), organizationId);
```
I remember we had `null` values for sourceId and email when testing.

jhaeyaert commented on gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/MongoUserRepository.java:
```diff
@@ -123,6 +138,16 @@ public User create(User user) throws TechnicalException {
         UserMongo userMongo = mapper.map(user);
+
+        if (isEncryptionEnabled) {
+            if (userMongo.getSourceId() != null) {
```
Indeed we had `null` values .😅

jhaeyaert commented on gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/upgrade/upgrader/index/__datakeys/DataKeysIndexUpgrader.java:
```diff
@@ -0,0 +1,45 @@
+            .name("k1")
+            .key("keyAltNames", ascending())
+            .unique(true)
+            .build();
```
For the record, we need to add the partialIndex as we saw it together in the [MongoDB doc](https://www.mongodb.com/docs/manual/core/csfle/quick-start/#create-a-unique-index-on-your-key-vault-collection).

phiz71 replied:
👍🏻 I have added the "partial" information to the index.

jhaeyaert commented on gravitee-apim-repository/gravitee-apim-repository-mongodb/src/test/java/io/gravitee/repository/mongodb/MongoTestRepositoryConfiguration.java:
```diff
@@ -119,4 +135,36 @@ private String getMongoFullVersion(String[] containerEnvs) {
+        Map<String, Object> localKmsProvider = new HashMap<>();
+        localKmsProvider.put(
+            "key",
+            "s1LSgcxwixnMWuPMWuNQpZZVHbaaT9pbyAuqrExk38gmcgg9DomK7IoYsvWpG+aZcOB7MfGTdRVILjNWL4mbEQ00f0Kz+W6f9Hu+npzOpMsbkH3SC/FlJEaeWrwavNQZ"
```
Can it be moved in the `graviteeTest.yaml` (if not too complicated)?

jhaeyaert commented on gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/encryption/EncryptionConfiguration.java:
```diff
@@ -0,0 +1,162 @@
+    @Value("${management.mongodb.encryption.keyVault.collectionName:__dataKeys}")
+    private String keyVaultCollectionName;
+
+    @Value("${management.mongodb.prefix}")
```
A default empty value is needed, ':' must be added, e.g.

```
    @Value("${management.mongodb.prefix:}")
    private String tablePrefix;
```

jhaeyaert commented on gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/internal/model/MetadataMongo.java:
```diff
@@ -41,5 +43,6 @@ public class MetadataMongo extends Auditable {
     private String format;
 
+    @ExplicitEncrypted(algorithm = AEAD_AES_256_CBC_HMAC_SHA_512_Deterministic)
     private String value;
```
Unrelated to encryption but after testing the metadata encryption I've noticed the updated date is never updated (there is a field mapping missing in the `MongoMetadataRepository.update()` method.

jhaeyaert replied:
The encryption of the metadata value works well by the way 🙂

phiz71 committed:
fix(mongodb): update metadata updatedAt field

phiz71 committed:
feat: encrypt sensitive data

License and Token

phiz71 committed:
feat: add extra jar in dev docker images

For Mongodb encryption to work, we need additional libraries. One of them is nearly 30Mb. As it is not yet an official support, we have decided to add them manually in the dev docker image only, instead of adding them in the mongodb repository plugin.

phiz71 committed:
feat: move users sort in service layer

Encryption has been introduced in mongo repository plugin.
As a consequence, the search method for users has to rely on encrypted field to be consistent.
To order users by `name` (or another encrypted field), Lucene search engine has to be used
(See io.gravitee.rest.api.service.impl.UserServiceImpl.search(ExecutionContext, String, Pageable))

ThibaudAV commented on gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/internal/user/UserMongoRepository.java:
```diff
@@ -34,9 +34,15 @@ public interface UserMongoRepository extends MongoRepository<UserMongo, String>,
 
     @Query(value = "{ 'source': ?0, 'sourceId': {$regex: '^?1$', $options: 'i'}, 'organizationId': ?2 }")
-    UserMongo findBySourceAndSourceId(String source, String sourceId, String organizationId);
+    UserMongo findBySourceAndSourceIdIgnoreCase(String source, String sourceId, String organizationId);
 
     @Query(value = "{ 'email': {$regex: '^?0$', $options: 'i'}, 'organizationId': ?1 }")
+    UserMongo findByEmailIgnoreCase(String email, String organizationId);
+
+    @Query(value = "{ 'source': ?0, 'sourceId': ?1, 'organizationId': ?2 }")
+    UserMongo findBySourceAndSourceId(String source, String sourceId, String organizationId);
+
+    @Query(value = "{ 'email':?0, 'organizationId': ?1 }")
```
Juste a remind : 
I don't know if in the database the email or sourceId have a value with capital letters in the database. If this is the case, they will need to be passed in lowercase with the encprition of the data. and remember to test it.
right ?

phiz71 replied:
ToLowercase has been added in the migration script and successfully tested
